[FSDP][optim_state_dict][6/N] Refactor the optim_state_dict APIs to support hooks

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91234
* #91160
* __->__ #90798

**What does this PR do?**

This PR splits the FSDP optim_state_dict APIs into common implementation parts that are shared for different frontend APIs (we have many now and will consolidate them gradually). This PR also add `_optim_state_dict_post_hook` and `_load_optim_state_dict_pre_hook` for the integration with `NamedOptimzer`.

fegin committed:
[FSDP][optim_state_dict][6/N] Refactor the optim_state_dict APIs to support hooks

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][6/N] Refactor the optim_state_dict APIs to support hooks"




[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1129,6 +1129,102 @@ def _raise_on_use_orig_params_optim_checkpoint(
+            True,
+            use_orig_params,
+        )
+        if is_named_optimizer:
```
might be useful to have a small comment here saying that NamedOptim expects the keys to be FQNs instead of regular optimizers.

fegin replied:
Will add this in the next PR.

rohan-varma reviewed (approved):
LGTM, thanks!

fegin commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1204,6 +1204,7 @@ def _unflatten_process_groups(
 def _optim_state_dict(
     model: torch.nn.Module,
     optim: torch.optim.Optimizer,
+    optim_state_dict: Dict[str, Any],
```
Both are accepted.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1129,6 +1129,102 @@ def _raise_on_use_orig_params_optim_checkpoint(
+
+        use_orig_params = False
+        for module in FullyShardedDataParallel.fsdp_modules(model):
+            use_orig_params = module._use_orig_params
```
Sure, can error out if that's not true.

awgu reviewed (approved):
Nice!

fegin committed:
Update on "[FSDP][optim_state_dict][6/N] Refactor the optim_state_dict APIs to support hooks"


**What does this PR do?**

This PR splits the FSDP optim_state_dict APIs into common implementation parts that are shared for different frontend APIs (we have many now and will consolidate them gradually). This PR also add `_optim_state_dict_post_hook` and `_load_optim_state_dict_pre_hook` for the integration with `NamedOptimzer`.



[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[FSDP][optim_state_dict][6/N] Refactor the optim_state_dict APIs to support hooks"


**What does this PR do?**

This PR splits the FSDP optim_state_dict APIs into common implementation parts that are shared for different frontend APIs (we have many now and will consolidate them gradually). This PR also add `_optim_state_dict_post_hook` and `_load_optim_state_dict_pre_hook` for the integration with `NamedOptimzer`.



[ghstack-poisoned]

fegin commented:
@pytorchbo merge

fegin commented:
@pytorchbot merge

fegin commented:
@pytorchbot merge -f "The failing test is not related"
